src/hammer-vlsi/hammer_vlsi/constraints.py

```diff
@@ -80,19 +82,153 @@ def from_setting(d: dict) -> "SRAMParameters":
+                    "Pins {p} have invalid side {s}. Assuming pins will be handled by CAD tool.".format(p=pins,
+                                                                                                        s=raw_side))
+
+        preplaced = raw_assign.get("preplaced", False)  # type: bool
+        assert isinstance(preplaced, bool), "preplaced must be a bool"
+
+        location = None  # type: Optional[Tuple[float, float]]
+        if "location" in raw_assign:
+            location_raw = raw_assign["location"]  # type: Union[List[float], Tuple[float, float]]
+            assert len(location_raw) == 2, "location must be a Optional[Tuple[float, float]]"
+            assert isinstance(location_raw[0], float), "location must be a Optional[Tuple[float, float]]"
+            assert isinstance(location_raw[1], float), "location must be a Optional[Tuple[float, float]]"
+            location = (location_raw[0], location_raw[1])
+        if not semi_auto and location is not None:
+            raise PinAssignmentSemiAutoError("location requires semi_auto")
+
+        width = raw_assign.get("width", None)  # type: Optional[float]
+        if width is not None:
+            assert isinstance(width, float), "width must be a float"
+        if not semi_auto and width is not None:
+            raise PinAssignmentSemiAutoError("width requires semi_auto")
+
+        depth = raw_assign.get("depth", None)  # type: Optional[float]
+        if depth is not None:
+            assert isinstance(depth, float), "depth must be a float"
+        if not semi_auto and depth is not None:
+            raise PinAssignmentSemiAutoError("depth requires semi_auto")
+
+        layers = None  # type: Optional[List[str]]
+        if "layers" in raw_assign:
+            raw_layers = raw_assign["layers"]  # type: List[str]
+            assert isinstance(raw_layers, list), "layers must be a List[str]"
+            for layer in "layers":
+                assert isinstance(layer, str), "layers must be a List[str]"
+            layers = raw_layers
+
+        if preplaced:
+            should_be_none = reduce(operator.and_, map(lambda x: x is None, [side, location, width, depth]))
+            if len(get_or_else(layers, cast(List[str], []))) > 0 or not should_be_none:
+                raise PinAssignmentPreplacedError(
+                    PinAssignment(pins=pins, side=None, layers=[], preplaced=preplaced, location=None, width=None,
+                                  depth=None))
+        else:
+            if len(get_or_else(layers, cast(List[str], []))) == 0 or side is None:
+                raise PinAssignmentError(
+                    "Pins {p} assigned without layers or side. Assuming pins will be handled by CAD tool.".format(
+                        p=pins))
+        return PinAssignment(
+            pins=pins,
+            side=side,
+            layers=layers,
+            preplaced=preplaced if "preplaced" in raw_assign else None,
```

Comment:
We want the default to be `Some(False)`.

Reply:
Hmm, maybe it might be necessary to store an underlying preplaced value that's not the default. For example, if we just put False here, it breaks a round trip of JSON->object->JSON comparison as it injects the default instead of preserving the None.

Reply:
I think thats kind of the point. We don't really want this to be optional(we're just letting it be so for the lazy users see: https://en.wikipedia.org/wiki/Robustness_principle), so when hammer emits it hammer will always use the default.

Reply:
Maybe in that case we should just set the internal type of `preplaced` to bool then. I can go ahead and do that

Reply:
I mean thats not what I want. I like not having to say preplaced in my input yamls. I just think we should be strict in what we output. Because the output is computer generated while the input is often human generated.

Reply:
I think we might be talking about different things here; I'd say to allow the dict parser to have an optional `preplaced` but have the internal Python type be just `bool` so that import/export is consistent.

Reply:
Ah ok. Yes that makes sense.